### PR TITLE
1123 open ai mode

### PR DESCRIPTION
**Pull request by fwild**
implements #1123

### CODE REVIEW

**Comment by fwild** [comment]
@cawke can you please review this? I was struggling with the movement of the Library files (stupidly, I also had moved them in this branch, before you did the changes on develop), but it is now all clean - and the functionality works, ready to be tested!

File: Assets/MirageXR/Services/IBM-Watson/DialogueService.cs [hunk 1]
```diff
@@ -75,108 +87,137 @@ void Start()
...
-        Credentials credentials = null;
 
-        //  Authenticate using iamApikey
-        TokenOptions tokenOptions = new TokenOptions()
+        if (AI == AIservice.openAI)
         {
-            IamApiKey = iamApikey
-        };
+            _openAIinterface = new OpenAI_API.OpenAIAPI();
+            AppLog.Log("[DialogueService] setting AI provider to openAI for organization = '" + _openAIinterface.Auth.OpenAIOrganization + "' and key='" + _openAIinterface.Auth.ApiKey + "'", LogLevel.INFO);
```
Comment (cawke) [inline Assets/MirageXR/Services/IBM-Watson/DialogueService.cs]: pleace add NRE check

File: Assets/MirageXR/Services/IBM-Watson/DialogueService.cs [hunk 2]
```diff
@@ -75,108 +87,137 @@ void Start()
 
     }
 
-    private IEnumerator CreateService()
+    public IEnumerator CreateService()
```
Comment (cawke) [inline Assets/MirageXR/Services/IBM-Watson/DialogueService.cs]: split it into two different methods (for openAI and IBMWatson)

File: Assets/MirageXR/ContentTypes/Character/Scripts/CharacterController.cs [hunk 3]
```diff
@@ -444,6 +463,39 @@ private async void AddWatsonAssistant()
...
+
+            // switch provider if needed
+            var dialogueService = watsonService.transform.Find("WatsonServices").GetComponent<DialogueService>();
+            if (!useWatson)
+            {
+                dialogueService.AI = AIservice.openAI;
+                _useWatson = false;
+                Runnable.Run(dialogueService.CreateService());
+            }
+            else if (useWatson)
```
Comment (cawke) [inline Assets/MirageXR/ContentTypes/Character/Scripts/CharacterController.cs]:
```suggestion
            else
```

**Review by cawke — CHANGES_REQUESTED** [review]
in fact, I don’t quite understand how the character should be controlled and what he should be able to do, but now if you tell him something, text will simply appear above him and he will not react to subsequent phrases.

image

**Comment by cawke** [comment]
Also, errors often appear when switching steps if there is a character in one of them

File: Assets/MirageXR/Services/IBM-Watson/DialogueService.cs (hunk 2 above)
Comment (fwild) [reply Assets/MirageXR/Services/IBM-Watson/DialogueService.cs]: I think we need to completely rewrite the whole structure - I don't think we want to keep it like this, it's even mixing two styles...

**Comment by fwild** [comment]
> Also, errors often appear when switching steps if there is a character in one of them

I have checked this but could not directly see errors?

**Comment by fwild** [comment]
I think before we merge, I'd want to see if iOS and Android build ok? Could you check that at a time convenient, Alex?

**Comment by cawke** [comment]
> I think before we merge, I'd want to see if iOS and Android build ok? Could you check that at a time convenient, Alex?

https://mirage-community.slack.com/files/U016VTVG29F/F06BM9DQGLW/build_openai.apk

**Pull request merged**